File: opus/claude_opus/third_002_naive.js
const express = require('express');
const crypto = require('crypto');
const router = express.Router();

// In-memory order store (replace with your database layer)
const orders = new Map();

// Secret shared with the payment provider for signature verification
const WEBHOOK_SECRET = process.env.PAYMENT_WEBHOOK_SECRET || '';

// Track processed event IDs to ensure idempotency (use a persistent store in production)
const processedEvents = new Set();

/**
 * Verify the webhook signature using HMAC-SHA256.
 * Uses a timing-safe comparison to prevent timing attacks.
 */
function verifySignature(rawBody, signatureHeader) {
  if (!WEBHOOK_SECRET || !signatureHeader) {
    return false;
  }

  const expected = crypto
    .createHmac('sha256', WEBHOOK_SECRET)
    .update(rawBody)
    .digest('hex');

  const expectedBuf = Buffer.from(expected, 'utf8');
  const receivedBuf = Buffer.from(signatureHeader, 'utf8');

  if (expectedBuf.length !== receivedBuf.length) {
    return false;
  }

  return crypto.timingSafeEqual(expectedBuf, receivedBuf);
}

/**
 * Map a payment provider event type to an internal order status.
 */
function mapEventToStatus(eventType) {
  const statusMap = {
    'payment.succeeded': 'paid',
    'payment.failed': 'payment_failed',
    'payment.refunded': 'refunded',
    'payment.disputed': 'disputed',
    'payment.canceled': 'canceled',
    'payment.pending': 'pending',
  };
  return statusMap[eventType] || null;
}

/**
 * Webhook endpoint.
 * NOTE: This route requires the raw request body for signature verification.
 * Register it BEFORE any global express.json() body parser, e.g.:
 *
 *   app.use('/webhooks/payment',
 *     express.raw({ type: 'application/json' }),
 *     paymentWebhookRouter);
 */
router.post('/', (req, res) => {
  try {
    // req.body is a Buffer when using express.raw()
    const rawBody = Buffer.isBuffer(req.body)
      ? req.body
      : Buffer.from(JSON.stringify(req.body));

    const signature = req.get('X-Webhook-Signature') || req.get('X-Signature');

    if (!verifySignature(rawBody, signature)) {
      return res.status(401).json({ error: 'Invalid signature' });
    }

    let event;
    try {
      event = JSON.parse(rawBody.toString('utf8'));
    } catch (parseErr) {
      return res.status(400).json({ error: 'Invalid JSON payload' });
    }

    const eventId = event.id;
    const eventType = event.type;
    const data = event.data || {};
    const orderId = data.orderId || (data.object && data.object.orderId);

    if (!eventId || !eventType || !orderId) {
      return res.status(400).json({ error: 'Missing required event fields' });
    }

    // Idempotency: acknowledge duplicates without reprocessing
    if (processedEvents.has(eventId)) {
      return res.status(200).json({ received: true, duplicate: true });
    }

    const newStatus = mapEventToStatus(eventType);
    if (!newStatus) {
      // Unknown but valid event — acknowledge to prevent retries
      processedEvents.add(eventId);
      return res.status(200).json({ received: true, ignored: true });
    }

    const order = orders.get(orderId);
    if (!order) {
      return res.status(404).json({ error: 'Order not found' });
    }

    // Update the order status
    order.status = newStatus;
    order.updatedAt = new Date().toISOString();
    order.lastPaymentEvent = {
      id: eventId,
      type: eventType,
      processedAt: order.updatedAt,
    };
    orders.set(orderId, order);

    processedEvents.add(eventId);

    return res.status(200).json({
      received: true,
      orderId,
      status: newStatus,
    });
  } catch (err) {
    console.error('Webhook processing error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

module.exports = router;

File: firstrun/devstral-glm/glm_4_7_flash/third_002_naive.py
from flask import Flask, request, jsonify
import hmac
import hashlib
import json

app = Flask(__name__)

# In a production environment, load this from environment variables
WEBHOOK_SECRET = "your-secret-key-here"

# Mock database to simulate order storage
orders_db = {}

def verify_webhook_signature(payload: str, signature_header: str) -> bool:
    """
    Verifies the HMAC signature of the incoming webhook payload.
    """
    if not signature_header:
        return False
    
    # Calculate the expected signature
    expected_signature = hmac.new(
        WEBHOOK_SECRET.encode('utf-8'),
        payload.encode('utf-8'),
        hashlib.sha256
    ).hexdigest()
    
    # Use constant-time comparison to prevent timing attacks
    return hmac.compare_digest(expected_signature, signature_header)

@app.route('/webhook', methods=['POST'])
def handle_webhook():
    # 1. Get the raw payload
    payload = request.get_data(as_text=True)

    # 2. Get the signature from headers (e.g., X-Signature)
    signature = request.headers.get('X-Signature')

    # 3. Verify the signature
    if not verify_webhook_signature(payload, signature):
        return jsonify({"error": "Invalid signature"}), 401

    try:
        # 4. Parse JSON payload
        event_data = json.loads(payload)
        event_type = event_data.get('event')
        order_details = event_data.get('data')

        if not order_details:
            return jsonify({"error": "Missing order data"}), 400

        order_id = order_details.get('order_id')

        # 5. Update Order Status based on Event Type
        if event_type == 'payment.succeeded':
            orders_db[order_id] = {'status': 'paid', 'updated_at': 'now'}
        elif event_type == 'payment.failed':
            orders_db[order_id] = {'status': 'failed', 'updated_at': 'now'}
        elif event_type == 'payment.refunded':
            orders_db[order_id] = {'status': 'refunded', 'updated_at': 'now'}
        else:
            # Log unknown event type or ignore
            pass

        return jsonify({"status": "success", "order_id": order_id}), 200

    except json.JSONDecodeError:
        return jsonify({"error": "Invalid JSON"}), 400
    except Exception as e:
        return jsonify({"error": str(e)}), 500

if __name__ == '__main__':
    app.run(port=5000)